### PR TITLE
#354 Displaying unavailable time of team members fixed

### PR DESCRIPTION
**Pull request by YuraRiabov**
Screenshot_847
Screenshot_848

### CODE REVIEW

**Review by olena-kozak — APPROVED** [review]

File: frontend/src/app/modules/new-meeting/new-meeting-calendar/new-meeting-calendar.component.ts [hunk 1]
```diff
@@ -108,17 +108,70 @@ export class NewMeetingCalendarComponent extends BaseComponent {
...
-            (this.containsDate(start, date, dateEnd) && this.containsDate(end, date, dateEnd));
+        return this.containsDate(date, unavailability.start, unavailability.end, unstrict) ||
+            this.containsDate(dateEnd, unavailability.start, unavailability.end, unstrict) ||
+            (this.containsDate(unavailability.start, date, dateEnd, true) &&
+                this.containsDate(unavailability.end, date, dateEnd, true));
     }
 
-    private containsDate(date: Date, start: Date, end: Date) {
+    private containsDate(date: Date, start: Date, end: Date, unstrict?: boolean) {
+        if (unstrict === true) {
```
Comment (tatianahutii) [inline frontend/src/app/modules/new-meeting/new-meeting-calendar/new-meeting-calendar.component.ts]:
```suggestion
        if (unstrict) {
```

File: frontend/src/app/modules/new-meeting/new-meeting-calendar/new-meeting-calendar.component.ts [hunk 2]
```diff
@@ -108,17 +108,70 @@ export class NewMeetingCalendarComponent extends BaseComponent {
...
-            this.containsDate(dateEnd, start, end) ||
-            (this.containsDate(start, date, dateEnd) && this.containsDate(end, date, dateEnd));
+        return this.containsDate(date, unavailability.start, unavailability.end, unstrict) ||
+            this.containsDate(dateEnd, unavailability.start, unavailability.end, unstrict) ||
+            (this.containsDate(unavailability.start, date, dateEnd, true) &&
+                this.containsDate(unavailability.end, date, dateEnd, true));
     }
 
-    private containsDate(date: Date, start: Date, end: Date) {
+    private containsDate(date: Date, start: Date, end: Date, unstrict?: boolean) {
```
Comment (tatianahutii) [inline frontend/src/app/modules/new-meeting/new-meeting-calendar/new-meeting-calendar.component.ts]: Move to the date helper

File: frontend/src/app/modules/new-meeting/new-meeting-calendar/new-meeting-calendar.component.ts [hunk 3]
```diff
@@ -108,17 +108,70 @@ export class NewMeetingCalendarComponent extends BaseComponent {
...
+    private isUnavailable(date: Date, unavailability: IUnavailability, unstrict?: boolean) {
         const dateEnd = addMinutes(date, this.duration.minutes!);
 
-        return this.containsDate(date, start, end) ||
-            this.containsDate(dateEnd, start, end) ||
-            (this.containsDate(start, date, dateEnd) && this.containsDate(end, date, dateEnd));
+        return this.containsDate(date, unavailability.start, unavailability.end, unstrict) ||
+            this.containsDate(dateEnd, unavailability.start, unavailability.end, unstrict) ||
+            (this.containsDate(unavailability.start, date, dateEnd, true) &&
+                this.containsDate(unavailability.end, date, dateEnd, true));
```
Comment (tatianahutii) [inline frontend/src/app/modules/new-meeting/new-meeting-calendar/new-meeting-calendar.component.ts]: Add two different methods and use them

File: frontend/src/app/modules/new-meeting/new-meeting-calendar/new-meeting-calendar.component.ts [hunk 4]
```diff
@@ -108,17 +108,70 @@ export class NewMeetingCalendarComponent extends BaseComponent {
...
+            const currentIntercepting = array.filter(e => this.isUnavailable(e.start, value, true) ||
+                this.isUnavailable(new Date(e.end), value, true));
+
+            const mergedEvent: IUnavailability = {
+                start: currentIntercepting.reduce((p, c) => (p.start < c.start ? p : c)).start,
+                end: currentIntercepting.reduce((p, c) => (p.end > c.end ? p : c)).end,
+            };
+
+            if (!merged.find(e => e.start === mergedEvent.start && e.end === mergedEvent.end)) {
+                merged.push(mergedEvent);
```
Comment (tatianahutii) [inline frontend/src/app/modules/new-meeting/new-meeting-calendar/new-meeting-calendar.component.ts]: Why push?

File: frontend/src/app/modules/new-meeting/new-meeting-calendar/new-meeting-calendar.component.ts [hunk 5]
```diff
@@ -108,17 +108,70 @@ export class NewMeetingCalendarComponent extends BaseComponent {
...
+    private mergeUnavailabilities(events: IUnavailability[]) {
+        const merged: IUnavailability[] = [];
+
+        events.map((value, index, array) => {
+            const currentIntercepting = array.filter(e => this.isUnavailable(e.start, value, true) ||
+                this.isUnavailable(new Date(e.end), value, true));
+
+            const mergedEvent: IUnavailability = {
+                start: currentIntercepting.reduce((p, c) => (p.start < c.start ? p : c)).start,
+                end: currentIntercepting.reduce((p, c) => (p.end > c.end ? p : c)).end,
```
Comment (tatianahutii) [inline frontend/src/app/modules/new-meeting/new-meeting-calendar/new-meeting-calendar.component.ts]: create a separate method for reduce

File: frontend/src/app/modules/new-meeting/new-meeting-calendar/new-meeting-calendar.component.ts [hunk 6]
```diff
@@ -108,17 +108,70 @@ export class NewMeetingCalendarComponent extends BaseComponent {
...
+            }
+
+            return e;
+        });
+    }
+
+    private mergeUnavailabilities(events: IUnavailability[]) {
+        const merged: IUnavailability[] = [];
+
+        events.map((value, index, array) => {
```
Comment (tatianahutii) [inline frontend/src/app/modules/new-meeting/new-meeting-calendar/new-meeting-calendar.component.ts]: Wrong using of map

**Review by tatianahutii — CHANGES_REQUESTED** [review]

**Review by tatianahutii — APPROVED** [review]

**Review by apofis96 — APPROVED** [review]

**Pull request merged**